packages/theme-language-server-common/src/json/JSONLanguageService.ts

```diff
@@ -0,0 +1,133 @@
+          ],
+        },
+      ],
+    });
+  }
+
+  async completions(params: CompletionParams): Promise<null | CompletionList | CompletionItem[]> {
+    if (!this.service) return null;
+    const documents = this.getDocuments(params);
+    if (!documents) return null;
+    const [jsonTextDocument, jsonDocument] = documents;
+    return this.service.doComplete(jsonTextDocument, params.position, jsonDocument);
+  }
+
+  async hover(params: HoverParams): Promise<Hover | null> {
+    if (!this.service) return null;
+    const documents = this.getDocuments(params);
+    if (!documents) return null;
+    const [jsonTextDocument, jsonDocument] = documents;
+    return this.service.doHover(jsonTextDocument, params.position, jsonDocument);
+  }
+
+  private getDocuments(
+    params: HoverParams | CompletionParams,
+  ): [TextDocument, JSONDocument] | null {
+    if (!this.service) return null;
+
+    const document = this.documentManager.get(params.textDocument.uri);
+    if (!document) return null;
+
+    switch (document.type) {
+      case SourceCodeType.JSON: {
+        const jsonTextDocument = document.textDocument;
+        const jsonDocument = this.service.parseJSONDocument(jsonTextDocument);
+        return [jsonTextDocument, jsonDocument];
+      }
+
+      case SourceCodeType.LiquidHtml: {
+        if (document.ast instanceof Error) return null;
+        const textDocument = document.textDocument;
+        const offset = textDocument.offsetAt(params.position);
+        const [_, ancestors] = findCurrentNode(document.ast, offset);
+        const schema = ancestors.find(
+          (node): node is LiquidRawTag =>
+            node.type === NodeTypes.LiquidRawTag && node.name === 'schema',
+        );
+        if (!schema) return null;
+
+        const schemaLineNumber = textDocument.positionAt(schema.blockStartPosition.end).line;
+        // Hacking away "same line numbers" here by prefixing the file with newlines
+        // This way params.position will be at the same line number in this fake jsonTextDocument
+        // Which means that the completions will be at the same line number in the Liquid document
+        const jsonString =
+          Array(schemaLineNumber).fill('\n').join('') +
+          schema.source.slice(schema.blockStartPosition.end, schema.blockEndPosition.start);
+        const jsonTextDocument = TextDocument.create(
+          textDocument.uri,
+          'json',
+          textDocument.version,
+          jsonString,
+        );
+        const jsonDocument = this.service.parseJSONDocument(jsonTextDocument);
+        return [jsonTextDocument, jsonDocument];
+      }
+    }
+  }
+
+  private async getSchemaForURI(uri: string): Promise<string> {
+    switch (uri) {
+      case SectionSchemaURI:
+        return this.jsonValidationSet.sectionSchema();
+      case TranslationFileURI:
+        return this.jsonValidationSet.translationSchema();
+      default:
+        throw new Error(`No schema for ${uri}`);
```

Comment:
This branch only happens if someone manually puts a "$schema" in their JSON objects. Which I don't believe we do have in themes.